lib/wasix/src/state/linker.rs

```diff
@@ -0,0 +1,751 @@
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum LinkError {
+    #[error("Linker not initialized")]
+    NotInitialized,
+
+    // FIXME: support needed subsection, remove this error
+    #[error("The 'needed' subsection of dylink.0 is not supported yet")]
+    NeededSubsectionNotSupported,
+
+    #[error("Main module is missing a required import: {0}")]
+    MissingMainModuleImport(String),
+
+    #[error("Module compilation error: {0}")]
+    CompileError(#[from] CompileError),
+
+    #[error("Failed to instantiate module: {0}")]
+    InstantiationError(#[from] InstantiationError),
+
+    #[error("Memory allocation error: {0}")]
+    MemoryAllocationError(#[from] MemoryError),
+
+    #[error("Runtime error: {0}")]
+    TableAllocationError(RuntimeError),
+
+    #[error("File system error: {0}")]
+    FileSystemError(#[from] FsError),
+
+    #[error("Module is not a dynamic library")]
+    NotDynamicLibrary,
+
+    #[error("Failed to parse dylink.0 section: {0}")]
+    Dylink0SectionParseError(#[from] wasmparser::BinaryReaderError),
+
+    #[error("Bad known import: {0} of type {1:?}")]
+    BadImport(String, ExternType),
+
+    #[error("Import could not be satisfied because it's missing: {0}")]
+    MissingImport(String),
+
+    #[error(
+        "Import could not be satisfied because of type mismatch: {0}, expected {1:?}, found {2:?}"
+    )]
+    ImportTypeMismatch(String, ExternType, ExternType),
+
+    #[error("Failed to initialize instance: {0}")]
+    InitializationError(anyhow::Error),
+
+    #[error("Initialization function has invalid signature: {0}")]
+    InitFuncWithInvalidSignature(String),
+
+    #[error("Initialization function {0} failed to run: {1}")]
+    InitFunctionFailed(String, RuntimeError),
+}
+
+pub enum ResolvedExport {
+    Function(Function),
+
+    // Contains the offset of the global in memory, with memory_base accounted for
+    // See: https://github.com/WebAssembly/tool-conventions/blob/main/DynamicLinking.md#exports
+    Global(u64),
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ResolveError {
+    #[error("Invalid module handle")]
+    InvalidModuleHandle,
+
+    #[error("Missing export")]
+    MissingExport,
+
+    #[error("Invalid export type: {0:?}")]
+    InvalidExportType(ExternType),
+}
+
+pub struct DylinkInfo {
+    pub mem_info: wasmparser::MemInfo,
+}
+
+pub struct LinkedMainModule {
+    pub memory: Memory,
+    pub indirect_function_table: Table,
+    pub stack_low: u64,
+    pub stack_high: u64,
+}
+
+impl Linker {
+    // TODO: It makes more sense to move the entire instantiation flow here, but that requires a bigger
+    // refactor in WasiEnv::instantiate.
+    pub fn new_for_main_module(
+        main_module: &Module,
+        store: &mut impl AsStoreMut,
+        memory: Option<Memory>,
+        imports: &mut Imports,
+        stack_size: u64,
+    ) -> Result<(Self, LinkedMainModule), LinkError> {
+        let dylink_section = super::linker::parse_dylink0_section(&main_module)?;
```

Comment:
I know the linker is probably insecure about it's identity, but that function is still a part of it's own module!

Reply:
Lol, cut-pasted that from an external module 😂